### PR TITLE
cmake: Fix invalid/missing hex files in build folder

### PR DESCRIPTION
**Pull request by jonathannilsen**
**Prevent partition manager from creating empty hex files for container partitions without images (e.g. s1.hex when CONFIG_BUILD_S1_VARIANT=n)**
_When container partitions were present in pm.yml but did not have an associated hex file, partition manager created an empty hex file for that partition in the build directory._

**Change b0/b0n partition configurations so that image partition names match the child image names.**
_The bootloader/netboot child image names (b0/b0n) did not match the image partition names (b0_image/b0n_image).
This caused the merged hex files for the b0/b0n container partitions to not actually include the bootloader hex file, only the provisioning data.
By correcting the partition names, the bootloader hex is included in the merge._

Ref: NCSDK-5567
    
Signed-off-by: Jonathan Nilsen <Jonathan.Nilsen@nordicsemi.no>

### CODE REVIEW

**Comment by jonathannilsen** [comment]
Will try to find an alternative solution that does not change the names passed to add_child_image()

**Comment by tejlmand** [comment]
> Copy child image hex files before merging them, so that intermediate hex files are created in the application build directory.

Why ?
There should be no need for copying hex files around, let's use them where they are generated, that is in the build folder of the child image.

File: cmake/partition_manager.cmake [hunk 1]
```diff
@@ -263,9 +263,22 @@ foreach(part ${PM_ALL_BY_SIZE})
     list(APPEND explicitly_assigned ${part})
   else()
     if(${part} IN_LIST images)
-      set(${part}_PM_HEX_FILE ${${part}_ZEPHYR_BINARY_DIR}/${${part}_KERNEL_HEX_NAME})
+      set(${part}_PM_HEX_FILE ${PROJECT_BINARY_DIR}/${part}.hex)
       set(${part}_PM_ELF_FILE ${${part}_ZEPHYR_BINARY_DIR}/${${part}_KERNEL_ELF_NAME})
-      set(${part}_PM_TARGET ${part}_subimage)
+      set(${part}_PM_TARGET ${part}_hex)
+      add_custom_command(
```
Comment (tejlmand) [inline cmake/partition_manager.cmake:269]: Why is this needed ?
Also, please add some more description in the commit message.

I don't see how this commit relates to: NCSDK-5567
can you elaborate a bit.

File: samples/bootloader/pm.yml [hunk 2]
```diff
@@ -1,17 +1,17 @@
 #include <autoconf.h>
 
-b0_image:
+b0:
```
Comment (tejlmand) [inline samples/bootloader/pm.yml:3]: Nothing wrong here, but cannot place comments in the commit message.

Please add a bit more description on what/why this commit does in the commit message, and secondly, I don't think this commit should refer the NCSDK-5567, as it is just a cleanup commit, from what I can tell.

File: samples/nrf5340/netboot/pm.yml [hunk 3]
```diff
@@ -1,6 +1,6 @@
 #include <autoconf.h>
 
-b0n_image:
+b0n:
```
Comment (tejlmand) [inline samples/nrf5340/netboot/pm.yml:3]: consider squash this with previous commit as general cleanup commit.

File: subsys/bootloader/include/dummy_values/pm_config.h [hunk 4]
```diff
@@ -1,14 +1,14 @@
...
 #define PM_CONFIG_H__
-#define PM_B0_ADDRESS 0x0
-#define PM_B0_SIZE 0x8000
-#define PM_B0_IMAGE_ADDRESS 0x8000
-#define PM_B0_IMAGE_SIZE 0x1000
-#define PM_B0N_ADDRESS 0x0
-#define PM_B0N_SIZE 0x8000
-#define PM_B0N_IMAGE_ADDRESS 0x8000
-#define PM_B0N_IMAGE_SIZE 0x1000
+#define PM_B0_PROVISION_ADDRESS 0x0
```
Comment (tejlmand) [inline subsys/bootloader/include/dummy_values/pm_config.h:4]: squash with two previous commits.
I assume you cannot bisect over those three commit, as I assume the changes in this commit is required for successfully compiling the previous two commits.

**Review by tejlmand — CHANGES_REQUESTED** [review]

File: cmake/partition_manager.cmake (hunk 1 above)
Comment (jonathannilsen) [reply cmake/partition_manager.cmake:269]: It was my understanding from reading the issue that the intended behavior was that a copied hex file should be created, and that it was missing. I don't really have much justification apart from that, so I am happy to drop the change if it is not useful.

File: samples/bootloader/pm.yml (hunk 2 above)
Comment (jonathannilsen) [reply samples/bootloader/pm.yml:3]: I added a description to the commit and updated the PR message. The partition naming was the cause of the bug where an invalid merged b0.hex/b0n.hex was created, so I believe it is relevant to NCSDK-5567. The last two commits were just for cleanup, though.

File: cmake/partition_manager.cmake (hunk 1 above)
Comment (tejlmand) [reply cmake/partition_manager.cmake]: @oyvindronningstad can you join here ?
what would be the reason for copying the file.

File: samples/bootloader/pm.yml (hunk 2 above)
Comment (tejlmand) [reply samples/bootloader/pm.yml:3]: > The partition naming was the cause of the bug where an invalid merged b0.hex/b0n.hex was created

thanks for clarifying.
With the updated commit message, this is much clearer.

> The last two commits were just for cleanup, though.

there is nothing wrong in adding cleanup commits to a PR.
It's highly appreciated to cleanup code as part of making other changes.

But to make it easier for reviewers to distinguish between commits that fixes issues and cleanup that are done because you are anyway making changes, then it's important to clearly identify the kind of commit in the commit message.

**Review by tejlmand — APPROVED** [review]

**Comment by kyberdin** [comment]
Received permission from @jonathannilsen to get the PR passing CI and merge-ready.

**Comment by kyberdin** [comment]
CI fails due to:

> The PR author must be in the 'nrfconnect' github organization or the PR must have the 'CI-trusted-author' label.

**Comment by kyberdin** [comment]
Added a commit to address the span/image name issue NCSDK-4199 that would become relevant when these changes are merged.

File: applications/nrf_desktop/configuration/nrf52840dk_nrf52840/pm_static_ZDebugB0.yml [hunk 5]
```diff
@@ -1,11 +1,11 @@
-b0:
+b0_provision:
```
Comment (pdunaj) [inline applications/nrf_desktop/configuration/nrf52840dk_nrf52840/pm_static_ZDebugB0.yml:1]: isn't the new naming somehow odd? Container is named after one of the elements.
b0_provision contains b0 and provision

Comment (kyberdin) [reply applications/nrf_desktop/configuration/nrf52840dk_nrf52840/pm_static_ZDebugB0.yml:1]: @jonathannilsen can probably better speak to the reasoning behind the name here. AFAIK, the name of this container isn't very important so long as it doesn't match the name of the child image to which it belongs (NCSDK-7982).

Since `b0_provision` is just a container partition, perhaps `b0_span` is clearer for its purpose in the configuration?

Comment (jonathannilsen) [reply applications/nrf_desktop/configuration/nrf52840dk_nrf52840/pm_static_ZDebugB0.yml:1]: I named it `b0_provision` since that seemed to fit with the naming pattern used for other container partitions in samples/bootloader/pm.yml. I agree that a name like `b0_span` is clearer.

**Review by FrancescoSer — APPROVED** [review]

**Review by tejlmand — CHANGES_REQUESTED** [review]
Please cleanup / squash the additional review commits into right commits

**Review by FrancescoSer — APPROVED** [review]

**Review by tejlmand — APPROVED** [review]
